### ISSUE TITLE
Unified XML for Visual Editor

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#787]
Revisiting the Visual Editor design, I see that the Domino Visual Editor project (#646) has its own set of intermediate XML, parsing logic, intemediate classes, and editor components.   I feel that this goes against the original design of Visual Editor, so I'd like to unify this code.

The goals are:
- Remove duplication in the parsing logic
- Most of the logic for NSF Exporter should not need to change (except for adding new elements or attributes that are required for a new export language).
- Avoid duplicated work for the mockup UI elements
- Adding a new language should ideally be as simple as writing a new exporter for an existing editor.
- During export, ignore or add placeholders for components and attributes that don't make sense for the target language.   The implementation status should be documented (i.e. supported, not supported, not yet implemented).

Here is a diagram to help clarify what I have in mind:
![image](https://user-images.githubusercontent.com/24532731/108908246-aec5e780-75e8-11eb-8260-4b80252ef8e7.png)

Here is how I envision the classes being modified:

Parser/POJO Classes:   https://github.com/prominic/VisualEditorConverterLib/tree/features/issue_646_ve_notes_domino_support/src/components/primeFaces
- Parse the unified XML into a single set of objects that will be used by the Visual Editor and language-specific exporters
- Move these to a generic package, say src/components/generic
- Move the "toCode" logic to PrimeFaces Exporter classes

Domino Parser/POJO classes:  https://github.com/prominic/VisualEditorConverterLib/tree/master/src/components/domino
- These will be removed in the unified design
- The additional fields should be merged into the appropriate Parser Classes above
- Create additional parser components as well, but these should be defined as generic UI components rather than components specific to Domino.   So, RichText would be fine, but not DominoField.
- Move the toCode logic to Domino Exporter classes

Visual Editor Components, including: [propertyEditors](https://github.com/prominic/MockupVisualEditor/tree/features/issue_646_ve_notes_domino_support/src/view/primeFaces/propertyEditors), [surfaceComponents](https://github.com/prominic/MockupVisualEditor/tree/features/issue_646_ve_notes_domino_support/src/view/primeFaces/surfaceComponents/components)
- This should be controls to create, edit, and organize structures of the POJO classes above.  
- The UI logic should not be dependent on the export language - the user is building a mockup of the interface, and the exporters should generate an interface as similar as possible to this design.
- One execption is that we may allow filters on the UI to hide features that are not applicable for the language.  This is meant to be a convenience tool so that the user does not waste their time on features that are not relevant to the target language.

Domino Visual Editor components, including: [propertyEditors](https://github.com/prominic/MockupVisualEditor/tree/features/issue_646_ve_notes_domino_support/src/view/domino/propertyEditors), [surfaceComponents](https://github.com/prominic/MockupVisualEditor/tree/features/issue_646_ve_notes_domino_support/src/view/domino/surfaceComponents/components)
- These will be removed in the unified design
- Move the additional required components and attributes to the unified Visual Editor components
- Split up DominoField in to the specialized fields used by the old Primefaces logic.

Exporter classes (New):
- Input is the POJO structure parsed above, output is a generated file based on this  code
- One set of Primefaces exporter classes based on the existing toCode code
- One set of Domino exporter classes based on the existing toCode code

TODO:
- Unified project structure, so that NSF Converter doesn't need to generate additional structures

I included @JustinProminic @piotrzarzycki21, and @rat-moonshine on this issue to discuss the design further.   We can discuss this tomorrow in the meeting.

Let me know if you have questions or concerns about this design.   I see that it is a major refactoring, but I'm hoping we can reuse most of the code we have written.  In particular, please report:
- Additional classes that will need significant refactoring for this design
- Potential side effects or problems that I have not mentioned

**Comment by piotrzarzycki21** [comment]
@JoelProminic I'm not sure if that issue shouldn't be placed in VisualEditor repository. In the end Moonshine is just using it nothing more...

**Comment by piotrzarzycki21** [comment]
All the changes which are going to happen should be made I think after merge of @feather812002 branch - we can discuss that during our today's call

**Comment by piotrzarzycki21** [comment]
I think whatever we will do it should be covered by unit tests. There are present for current logic in ->https://github.com/prominic/VisualEditorConverterLib - When I checked some time ago they are still working nicely.

**Comment by piotrzarzycki21** [comment]
I think it will take quite a bit of time to do that, but I believe this refactoring is needed. We are growing with that and what was ok for a simple test cases at the beginning now may end up with bigger issues if we continue using old logic.

**Comment by feather812002** [comment]
I think refactor is fine and the design also is fine , I just want point out ,it need refactoring almost all domino visual code class.
because ,the intermediary xml format already become to  unified XML .
In my last year work , I build the domino  intermediary xml format ,it can support some special attributes for domino design element and it only work for domino design element. I already done this ,and it work fine with some domino design element.

- Parse the unified XML into a single set of objects that will be used by the Visual Editor and language-specific exporters

> Now ,we try to work on the unified xml, so , first thing ,I need rewrite the xml parse logic .and I can't understand what is the 
> **" a single set of objects "**, for current logic , the primeface xml format convert to primeface Object with "fromXML" by Converter. the domino xml format convert to domino Object with "fromXML" by DominoConvert.
> So, what meant about the **" a single set of objects "**,we need a clear defined for this in program level, rather than just some word on this.

- Move the "toCode" logic to PrimeFaces Exporter classes
> What is the  "PrimeFaces Exporter classes"? if we just simple move the "toCode" code to other place of the project ,I think this is easy , it just change some class name or file name . if I misunderstand this ,please let me know.

- These will be removed in the unified design

> This is okay.

- The additional fields should be merged into the appropriate Parser Classes above

> Can you details this ? how can we merged into the appropriate Parser Classes above?If you mean just change the class into other place or put them into other package , this is fine. if you mean merged in logic level, please clear defined which Parser Classes should be merged with these class.

- Create additional parser components as well, but these should be defined as generic UI components rather than components specific to Domino. So, RichText would be fine, but not DominoField.
  

> This is big trouble from my understand , I don't know how to write a generic UI components  can let it work with other UI components, for example , the domino design element support formual language ,it only work with domino design element .
> So ,how can I let the  generic UI components support for this ?

- Move the toCode logic to Domino Exporter classes
What is the "Domino Exporter classes" ? where can I found it ? or is it need me write it in next some time ?

**Comment by JoelProminic** [comment]
We discussed this in the meeting today.  We mostly discussed clarification of the design, but here are some additional notes:
- @piotrzarzycki21 is planning to make some example classes to explore how the unified solution will work.
- We are discussing whether we should merge #646 into master before starting on this major refactoring.   
- We are also discussing how we can do this refactoring in an incremental way. 

@JustinProminic reminded us that current goal for VisualEditor is:
1. The user selects and exports a database to the intermediate XML
2. Open the project in Moonshine and make some modifications using Visual Editor
3. Export a Royale interface and some Java that implement a simple CRUD interface.

I'd like to have the Unified XML implemented for the Royale and Java exporter, so that we don't have a third set of XML objects and UI components.  It may make more sense to work on the unified design while we are implementing the new exporter, but I'd want to work with the other export languages so that we don't make another specialized XML/class design.

@feather812002 suggested that we could have Moonshine convert the current intermediate XML to specialized XML for PrimeFaces or Domino, so that we could quickly have a proof of concept of a process like this:
1. Export an NSF to intermediate XML
2. Open in Moonshine and make changes with Visual Editor
3. Export to an NSF.  We can compare the resulting forms to the original forms to see how well we are preserving the original design

I'll talk more with @JustinProminic about what we would like to focus on for this in the short term.
